src/commands/buffer.rs

```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+            m.insert(Scope::new("source.swift").unwrap(), "// ");
+            m.insert(Scope::new("source.toml").unwrap(), "# ");
+            m.insert(Scope::new("source.yaml").unwrap(), "# ");
+            m.insert(Scope::new("text.tex").unwrap(), "% ");
+            m.insert(Scope::new("text.tex.latex").unwrap(), "% ");
+            m
+        };
+    }
+
+    let buffer = app.workspace.current_buffer().ok_or(BUFFER_MISSING)?;
+    let original_cursor = *buffer.cursor.clone();
+
+    let comment_content = {
+        let scope = buffer.syntax_definition.as_ref()
+            .map_or(Scope::new("none").unwrap(), |def| def.scope);
+
+        COMMENT_MAP.get(&scope).ok_or_else(|| {
+            format!("No comment syntax definition found for scope {}", scope)
+        })?
+    };
+
+    // Get the range of lines we'll comment based on
+    // either the current selection or cursor line.
+    let lines = match app.mode {
+        Mode::SelectLine(ref mode) => {
+            if mode.anchor >= buffer.cursor.line {
+                buffer.cursor.line..mode.anchor + 1
+            } else {
+                mode.anchor..buffer.cursor.line + 1
+            }
+        }
+        _ => buffer.cursor.line..buffer.cursor.line + 1,
+    };
+
+    let buffer_range = Range::new(
+        Position { line: lines.start, offset: 0 },
+        Position { line: lines.end, offset: 0 }
+    );
+
+    let buffer_range_content = buffer.read(&buffer_range).ok_or(CURRENT_LINE_MISSING)?;
+
+    // Split the buffer content on newlines and filter out any zero-length lines.
+    let lines: Vec<(usize, &str)> = lines
+        .zip(buffer_range_content.split("\n"))
+        .filter(|(_, line)| line.trim().len() > 0)
+        .collect();
+
+    // We look at all lines to see if they start with `comment_content` or not.
+    // If even a single line does not, we need to comment all lines out,
+    // otherwise remove `comment_content` on the start of each line.
+    let (toggle, offset) = lines.iter()
+        .map(|(_, line)| {
+            let content = line.trim_start();
+            (content.starts_with(comment_content), line.len() - content.len())
+        })
+        .fold((true, usize::MAX), |result, (toggle, offset)| {
+            (result.0 & toggle, result.1.min(offset))
+        });
+
+    // Move to the start of each of the line's content and
+    // insert/remove the comments, as a single operation.
+    buffer.start_operation_group();
+    if !toggle {
+        add_line_comment(buffer, &lines, offset, comment_content);
+    } else {
+        remove_line_comment(buffer, &lines, comment_content);
+    }
+    buffer.end_operation_group();
+
+    // Restore original cursor
+    buffer.cursor.move_to(original_cursor);
+
+    Ok(())
+}
+
+fn add_line_comment(buffer: &mut Buffer, lines: &[(usize, &str)], offset: usize, comment_content: &str) {
+    for (line, _) in lines {
+        let target = Position { line: *line, offset };
```

Comment:
```suggestion
    for (line_number, _) in lines {
        let target = Position { line: *line_number, offset };
```
I know it's at odds with the `Position` type's line/offset fields, but we're using `line` to represent _content_ in the function below, and it represents a number here, which is somewhat confusing. :slightly_smiling_face:

Reply:
Yeah, that's confusing, you're right. I'm gonna revise all of the naming 😅